### PR TITLE
For #18267: Updates preferences visibility depending on creditCardsStore

### PR DESCRIPTION
**Pull request by mcarare**
### Pull Request checklist

- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

**Comment by mcarare** [comment]
run-testDebugUnitTest stuck.

File: app/src/test/java/org/mozilla/fenix/settings/creditcards/CreditCardsSettingFragmentTest.kt [hunk 1]
```diff
@@ -0,0 +1,65 @@
+package org.mozilla.fenix.settings.creditcards
```
Comment (gabrielluong) [inline app/src/test/java/org/mozilla/fenix/settings/creditcards/CreditCardsSettingFragmentTest.kt:1]: Needs a license

File: app/src/main/java/org/mozilla/fenix/settings/creditcards/CreditCardsSettingFragment.kt [hunk 2]
```diff
@@ -23,10 +34,43 @@ import org.mozilla.fenix.settings.requirePreference
  */
 class CreditCardsSettingFragment : PreferenceFragmentCompat() {
 
+    private lateinit var creditCardsStore: CreditCardsFragmentStore
+    private var creditCardsAlreadyLoaded: Boolean = false
```
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/settings/creditcards/CreditCardsSettingFragment.kt:38]:
```suggestion
    private var isCreditCardsLoaded: Boolean = false
```

File: app/src/main/java/org/mozilla/fenix/settings/creditcards/CreditCardsSettingFragment.kt [hunk 3]
```diff
@@ -23,10 +34,43 @@ import org.mozilla.fenix.settings.requirePreference
...
     override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
         setPreferencesFromResource(R.xml.credit_cards_preferences, rootKey)
     }
 
+    override fun onCreateView(
+        inflater: LayoutInflater,
+        container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View? {
+        loadCreditCards()
```
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/settings/creditcards/CreditCardsSettingFragment.kt:57]: I don't fully understand a fragment lifecycle so this is more to help me satisfy my curious and help me understand this change. Can you explain why we need to `loadCreditCards()` here? Thanks

Comment (mcarare) [reply app/src/main/java/org/mozilla/fenix/settings/creditcards/CreditCardsSettingFragment.kt:57]: Yes, when navigating away from this fragment to the _add card_ or _manage cards_ fragment, this fragment is not destroyed so when we come back from them `onCreate` is not called anymore. The view tough is destroyed, so when we come back `onCreateView` is still called. Given the fact that in the add/manage screens user can modify the card list by adding or deleting, we need to reload the card list and use it to update the state and implicitly the view. That is the reason we have `loadCreditCards` in both places, and the boolean to check we do not load the same list twice, but still call it as early as possible.

File: app/src/main/java/org/mozilla/fenix/settings/creditcards/CreditCardsSettingFragment.kt [hunk 4]
```diff
@@ -74,4 +118,34 @@ class CreditCardsSettingFragment : PreferenceFragmentCompat() {
...
+        val manageSavedCardsPreference =
+            requirePreference<Preference>(R.string.pref_key_credit_cards_manage_saved_cards)
+        val addCreditCardsPreference =
+            requirePreference<Preference>(R.string.pref_key_credit_cards_add_credit_card)
+
+        manageSavedCardsPreference.isVisible = hasCreditCards
+        addCreditCardsPreference.isVisible = !hasCreditCards
+    }
+
+    private fun loadCreditCards() {
```
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/settings/creditcards/CreditCardsSettingFragment.kt:138]:
```suggestion
    /**
     * Fetches all the credit cards from the autofill storage and updates the
     * [CreditCardsFragmentStore] with the list of credit cards.
     */
    private fun loadCreditCards() {
```

File: app/src/main/java/org/mozilla/fenix/settings/creditcards/CreditCardsSettingFragment.kt [hunk 5]
```diff
@@ -74,4 +118,34 @@ class CreditCardsSettingFragment : PreferenceFragmentCompat() {
...
+            return
+        }
+
+        lifecycleScope.launch(Dispatchers.IO) {
+            val creditCards = requireComponents.core.autofillStorage.getAllCreditCards()
+            lifecycleScope.launch(Dispatchers.Main) {
+                creditCardsStore.dispatch(CreditCardsAction.UpdateCreditCards(creditCards))
+            }
+        }
+        isCreditCardsListLoaded = true
```
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/settings/creditcards/CreditCardsSettingFragment.kt:149]:
```suggestion

        isCreditCardsListLoaded = true
```

File: app/src/main/java/org/mozilla/fenix/settings/creditcards/CreditCardsSettingFragment.kt [hunk 6]
```diff
@@ -74,4 +118,34 @@ class CreditCardsSettingFragment : PreferenceFragmentCompat() {
...
+    private fun loadCreditCards() {
+        if (isCreditCardsListLoaded) {
+            return
+        }
+
+        lifecycleScope.launch(Dispatchers.IO) {
+            val creditCards = requireComponents.core.autofillStorage.getAllCreditCards()
+            lifecycleScope.launch(Dispatchers.Main) {
+                creditCardsStore.dispatch(CreditCardsAction.UpdateCreditCards(creditCards))
+            }
```
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/settings/creditcards/CreditCardsSettingFragment.kt:147]:
```suggestion

            lifecycleScope.launch(Dispatchers.Main) {
                creditCardsStore.dispatch(CreditCardsAction.UpdateCreditCards(creditCards))
            }
```

File: app/src/test/java/org/mozilla/fenix/settings/creditcards/CreditCardsSettingFragmentTest.kt [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        val activity = Robolectric.buildActivity(FragmentActivity::class.java).create().get()
+
+        activity.supportFragmentManager.beginTransaction()
+            .add(creditCardsSettingFragment, "CreditCardsSettingFragmentTest")
+            .commitNow()
+        testDispatcher.advanceUntilIdle()
+    }
+
+    @Test
+    fun `manage credit cards pref is visible if there are saved cards`() {
```
Comment (gabrielluong) [inline app/src/test/java/org/mozilla/fenix/settings/creditcards/CreditCardsSettingFragmentTest.kt:41]: Should we use GIVEN WHEN THEN for these tests?

```suggestion
    fun `GIVEN a list of saved credit cards THEN the "Manage Credit Card" preference is visible`() {
```

**Review by gabrielluong — APPROVED** [review]
r+ assuming the requested changes are done

File: app/src/test/java/org/mozilla/fenix/settings/creditcards/CreditCardsSettingFragmentTest.kt (hunk 7 above)
Comment (mcarare) [reply app/src/test/java/org/mozilla/fenix/settings/creditcards/CreditCardsSettingFragmentTest.kt:41]: Good ideea!

**Pull request merged**